src/script/interpreter.cpp

```diff
@@ -1259,6 +1431,63 @@ bool EvalScript(vector<vector<unsigned char> > &stack,
                 }
                 break;
 
+                case OP_CHECKDATASIG:
+                case OP_CHECKDATASIGVERIFY:
+                {
+                    // Make sure this remains an error before activation.
+                    if (!(flags & SCRIPT_ENABLE_CHECKDATASIG))
+                        return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+
+                    // (sig message pubkey -- bool)
+                    if (stack.size() < 3)
+                    {
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                    }
+
+                    valtype &vchSig = stacktop(-3);
+                    valtype &vchMessage = stacktop(-2);
+                    valtype &vchPubKey = stacktop(-1);
+
+                    if (!CheckDataSignatureEncoding(vchSig, flags, serror) ||
+                        !CheckPubKeyEncoding(vchPubKey, flags, serror))
+                    {
+                        // serror is set
+                        return false;
+                    }
+
+                    bool fSuccess = false;
+                    if (vchSig.size())
+                    {
+                        valtype vchHash(32);
+                        CSHA256().Write(vchMessage.data(), vchMessage.size()).Finalize(vchHash.data());
+                        uint256 messagehash(vchHash);
+
+                        CPubKey pubkey(vchPubKey);
+                        fSuccess = pubkey.Verify(messagehash, vchSig);
+                    }
+
+                    if (!fSuccess && (flags & SCRIPT_VERIFY_NULLFAIL) && vchSig.size())
+                    {
+                        return set_error(serror, SCRIPT_ERR_SIG_NULLFAIL);
+                    }
+
+                    popstack(stack);
+                    popstack(stack);
+                    popstack(stack);
+                    stack.push_back(fSuccess ? vchTrue : vchFalse);
+                    if (opcode == OP_CHECKDATASIGVERIFY)
+                    {
+                        if (fSuccess)
+                        {
+                            popstack(stack);
```

Comment:
I seems unnecessary to push true or false onto the stack only to pop it back off if VERIFY.  Why not use:

if (opcode == CDS) stack.push_back()
else if (!fSuccess) // if opcode == CDSV
  {
    return set_error(...)
  }